From a8f79648b4f1fccb2e4e60ec822f6ff9482f5eac Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Wed, 03 Mar 2021 16:02:50 +0200
Subject: [PATCH] [SW-20154] gaudi2_arc_gnu: check for min size of image

To prevent malicious accesses, the security recommendation
is to make sure all values need to be in suitable limits
before a subtraction happens. Hence check for the minimum
available image size before proceeding with signature
verification

Change-Id: I0ed20af0f55df88245cdb1b979faa4a37e7b669f
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/gaudi2_arc_gnu/secureboot.c b/gaudi2_arc_gnu/secureboot.c
index bd80ad1..009c1cb 100644
--- a/gaudi2_arc_gnu/secureboot.c
+++ b/gaudi2_arc_gnu/secureboot.c
@@ -175,11 +175,16 @@
 	uint32_t test_data_len = image_size;
 	unsigned char *test_data = (unsigned char *)(image_offset);
 	unsigned char *test_sig = test_data + test_data_len;
+	unsigned char *ptr_key_index, *ptr_key_prop, *test_public_key;
+
+	/* check if the image size has minimum sized contents */
+	if (test_data_len <= TOTAL_PUB_KEY_PROPS_SZ)
+		return SECUREBOOT_VERIFY_VER_FAILED;
 
 	/* Key header is located as the last 2 DWORDS of the image */
-	unsigned char *ptr_key_index = test_sig - KEY_PROP_HDR_SZ;
-	unsigned char *ptr_key_prop = test_sig - TOTAL_PUB_KEY_PROPS_SZ;
-	unsigned char *test_public_key =  ptr_key_prop;
+	ptr_key_index = test_sig - KEY_PROP_HDR_SZ;
+	ptr_key_prop = test_sig - TOTAL_PUB_KEY_PROPS_SZ;
+	test_public_key =  ptr_key_prop;
 
 	uint32_t fw_version = base_readl((uint32_t)test_data);
 
